src/main/java/org/testng/internal/ClassHelper.java

```diff
@@ -458,22 +392,81 @@ public static Object createInstance1(
-        Logger.getLogger(ClassHelper.class).error(e.getMessage(), e);
+    IParametersAnnotation parametersAnnotation =
+        finder.findAnnotation(constructor, IParametersAnnotation.class);
+    if (parametersAnnotation == null) {
+      // null if the annotation is @Factory
+      return null;
+    }
+    String[] parameterNames = parametersAnnotation.getValue();
+    Object[] parameters =
+        Parameters.createInstantiationParameters(
+            constructor,
+            "@Parameters",
+            finder,
+            parameterNames,
+            xmlTest.getAllParameters(),
+            xmlTest.getSuite());
+    return objectFactory.newInstance(constructor, parameters);
+  }
+
+
+  private static Object instantiateUsingDefaultConstructor(Class<?> declaringClass,
+      Map<Class<?>, IClass> classes, XmlTest xmlTest, IObjectFactory factory)
+      throws NoSuchMethodException, IllegalAccessException, InstantiationException {
+    // If this class is a (non-static) nested class, the constructor contains a hidden
+    // parameter of the type of the enclosing class
+    Class<?>[] parameterTypes = new Class[0];
+    Object[] parameters = new Object[0];
+    Class<?> ec = declaringClass.getEnclosingClass();
+    boolean isStatic = 0 != (declaringClass.getModifiers() & Modifier.STATIC);
+
+    // Only add the extra parameter if the nested class is not static
+    if ((null != ec) && !isStatic) {
+      parameterTypes = new Class[] {ec};
+      parameters = computeParameters(classes, ec, factory);
+    } // isStatic
+
+    Constructor<?> ct;
+    try {
+      ct = declaringClass.getDeclaredConstructor(parameterTypes);
+    } catch (NoSuchMethodException ex) {
+      ct = declaringClass.getDeclaredConstructor(String.class);
+      parameters = new Object[]{xmlTest.getName()};
+      // If ct == null here, we'll pass a null
+      // constructor to the factory and hope it can deal with it
+    }
+    return factory.newInstance(ct, parameters);
+  }
+
+  private static Object[] computeParameters(Map<Class<?>, IClass> classes,
+      Class<?> ec, IObjectFactory factory)
+      throws NoSuchMethodException, IllegalAccessException, InstantiationException {
+    // Create an instance of the enclosing class so we can instantiate
+    // the nested class (actually, we reuse the existing instance).
+    IClass enclosingIClass = classes.get(ec);
+    Object[] enclosingInstances;
+    if (null != enclosingIClass) {
+      enclosingInstances = enclosingIClass.getInstances(false);
+      if ((null == enclosingInstances) || (enclosingInstances.length == 0)) {
+        Object o = factory.newInstance(ec.getConstructor(ec));
+        enclosingIClass.addInstance(o);
+        enclosingInstances = new Object[]{o};
       }
+    } else {
+      enclosingInstances = new Object[]{ec.newInstance()};
     }
+    Object enclosingClassInstance = enclosingInstances[0];
 
-    return result;
+    return new Object[]{enclosingClassInstance};
```

Comment:
Why not returning an object as it is the intent of the method?

```java
if (null == enclosingIClass) {
  return ec.newInstance();
}
Object[] enclosingInstances = enclosingIClass.getInstances(false);
if (enclosingInstances != null && enclosingInstances.length > 0) {
  return enclosingInstances[0];
}

Object o = factory.newInstance(ec.getConstructor(ec));
enclosingIClass.addInstance(o);
return o;
```

Reply:
And adapt to where you use it:

```java
parameters = new Object[]{computeParameters(classes, ec, factory)};
```

Reply:
@juherr - That was a good one. I didn't see hard enough there... I have fixed it, amended the commit and pushed it. Can you please check now